gtsfm/utils/cycle_consistency.py

```diff
@@ -0,0 +1,233 @@
+
+        nodes_from_i1 = adj_list[i1]
+        nodes_from_i2 = adj_list[i2]
+
+        node_intersection = (nodes_from_i1).intersection(nodes_from_i2)
+
+        for node in node_intersection:
+            cycle_nodes = tuple(sorted([i1, i2, node]))
+            triplets.add(cycle_nodes)
+
+    return list(triplets)
+
+
+def compute_cycle_error(
+    i2Ri1_dict: Dict[Tuple[int, int], Rot3],
+    cycle_nodes: Tuple[int, int, int],
+    two_view_reports_dict: Dict[Tuple[int, int], TwoViewEstimationReport],
+    verbose: bool = True,
+) -> Tuple[float, Optional[float], Optional[float]]:
+    """Compute the cycle error by the magnitude of the axis-angle rotation after composing 3 rotations.
+
+    Note: i1 < i2 for every valid edge, by construction.
+
+    Args:
+        i2Ri1_dict: mapping from image pair indices to relative rotation.
+        cycle_nodes: 3-tuples of nodes that form a cycle. Nodes of are provided in sorted order.
+        two_view_reports_dict:
+        verbose: whether to dump to logger information about error in each Euler angle
+
+    Returns:
+        cycle_error: deviation from 3x3 identity matrix, in degrees. In other words,
+            it is defined as the magnitude of the axis-angle rotation of the composed transformations.
+        max_rot_error: maximum rotation error w.r.t. GT across triplet edges, in degrees.
+            If ground truth is not known for a scene, None will be returned instead.
+        max_trans_error: maximum translation error w.r.t. GT across triplet edges, in degrees.
+            If ground truth is not known for a scene, None will be returned instead.
+    """
+    cycle_nodes = list(cycle_nodes)
+    cycle_nodes.sort()
+
+    i0, i1, i2 = cycle_nodes
+
+    i1Ri0 = i2Ri1_dict[(i0, i1)]
+    i2Ri1 = i2Ri1_dict[(i1, i2)]
+    i0Ri2 = i2Ri1_dict[(i0, i2)].inverse()
+
+    # should compose to identity, with ideal measurements
+    i0Ri0 = i0Ri2.compose(i2Ri1).compose(i1Ri0)
+
+    I_3x3 = Rot3()
+    cycle_error = comp_utils.compute_relative_rotation_angle(I_3x3, i0Ri0)
+
+    # form 3 edges between fully connected subgraph (nodes i,j,k)
+    e_i = (i0, i1)
+    e_j = (i1, i2)
+    e_k = (i0, i2)
+
+    rot_errors = [two_view_reports_dict[e].R_error_deg for e in [e_i, e_j, e_k]]
+    trans_errors = [two_view_reports_dict[e].U_error_deg for e in [e_i, e_j, e_k]]
+
+    gt_known = all([err is not None for err in rot_errors])
+    if gt_known:
+        max_rot_error = np.max(rot_errors)
+        max_trans_error = np.max(trans_errors)
+    else:
+        # ground truth unknown, so cannot estimate error w.r.t. GT
+        max_rot_error = None
+        max_trans_error = None
+
+    if verbose:
+        i1Ri0_euler = Rotation.from_matrix(i1Ri0.matrix()).as_euler(seq="xyz", degrees=True).tolist()
+        i2Ri1_euler = Rotation.from_matrix(i2Ri1.matrix()).as_euler(seq="xyz", degrees=True).tolist()
+        i0Ri2_euler = Rotation.from_matrix(i0Ri2.matrix()).as_euler(seq="xyz", degrees=True).tolist()
```

Comment:
why cant we use gtsam functionality directly? https://github.com/borglab/gtsam/blob/6d6d129d05f64216cc8814dc7f24fcf05fd67f44/gtsam/geometry/Rot3.h#L464

Reply:
@akshay-krishnan would like to chat more about this. So here I'm assuming we have a camera with +z pointing out of the camera, and we rotate it by 30 degrees about the rotate axis:
```python
>>> Rotation.from_euler("y", 30, degrees=True).as_matrix()
array([[ 0.8660254,  0.       ,  0.5      ],
       [ 0.       ,  1.       ,  0.       ],
       [-0.5      ,  0.       ,  0.8660254]])
```

The equivalent in GTSAM is:

```python
>>> Rot3.Pitch(np.deg2rad(30))
R: [
	0.866025, 0, 0.5;
	0, 1, 0;
	-0.5, 0, 0.866025
]
```

But pitch only makes sense if it's a vehicle frame with vehicle pointing in the +x direction, with +y to the left, and +z upwards.

So I feel like using the term "Pitch" is less accurate.

Reply:
I guess there is
```python
>>> Rot3.Ry(np.deg2rad(30))
R: [
	0.866025, 0, 0.5;
	0, 1, 0;
	-0.5, 0, 0.866025
]
```

Reply:
I don't see a way to decode "Ry" though from a Rot3 instance, without calling `.Pitch()`, which I feel like isn't accurate. So I'm fine with using GTSAM for the constructor, but I think when decoding the euler angles, I prefer the scipy.spatial.transform

Reply:
.xyz()[1] ? 
https://github.com/borglab/gtsam/blob/6d6d129d05f64216cc8814dc7f24fcf05fd67f44/gtsam/geometry/Rot3.h#L464

Reply:
I dont like the idea of redundant calculation and memory transfer in a library, it's kinda acceptable in a unit test.

Reply:
Updated and removed the scipy code here